Cache-Hash-Data-Structure.adoc

```diff
@@ -0,0 +1,209 @@
+= Hash Data Structure
+
+Provide a Hash data structure to Infinispan users. The Hash is a flexible identifiable grouping of keys mapping to values.
+In this context, _flexible_ means it can have arbitrary mappings without needing a schema beforehand; _identifiable group_
+means the Hash structure receives a unique name and retrieves the complete key-value mapping.
+
+The Hash provides a general approach for users to represent their data structures using Infinispan. For example, by
+employing this, users can avoid using key prefixes to create an object in Cache.
+
+== Public API
+
+The work is for the new API only. The Hash data structure has the format of:
+
+[source,java]
+.Hash.java
+----
+public interface Hash<K, V> { // <1>
+
+   String getName();
+
+   V get(K key);
+
+   V put(K key, V value);
+
+   V remove(K key);
+
+   boolean containsKey(K key);
+
+   Set<K> keySet();
+
+   Collection<V> values();
+
+   Set<Map.Entry<K, V>> entrySet();
+
+   void clear();
+
+   int size();
+}
+----
+
+<1> The Hash data structure is a generic interface over `K` and `V`. We do not extend `Map<K, V>`, instead we add methods
+as needed.
+
+A `Hash<?, ?>` structure is local to a node. Meaning that operations applied to the local copy must be submitted
+through the public API before taking effect. After that, the `Hash<?, ?>` is distributed/replicated according to
+the configuration.
+
+We follow a state-based CRDT approach, merging the local Hash with the server's Hash. Since the `Hash<?, ?>` structures
+occupy a single entry in the Cache, the whole structure has a version instead of versioning per Hash property. Taking
+advantage of this, we can use the `compute*` method and directly apply the local changes. If a concurrent change happens,
+the `Hash<?, ?>` version is updated, and the replace fails, where the `compute*` retries again until it succeeds. Note
+this could lead to flooding the network, transmitting the `Hash<?, ?>` back and forth, but this only happens with heavy
+concurrent changes over the same structure.
+
+[IMPORTANT]
+.Inconsistency!
+====
+The use of `Hash<?, ?>` could return an outdated value, since the structure is a local copy.
+====
+
+To interact with the `Hash<K, V>`:
+
+[source, java]
+.HashCache.java
+----
+public interface HashCache<K> {
+
+   <HK, HV> HV put(K key, HK hashKey, HV hashValue);
+
+   <HK, HV> Hash<HK, HV> put(K key, Map<HK, HV> values);
+
+   <HK, HV> Map<HK, HV> get(K key, HK hashKey, HK ... others); // <1>
+
+   Hash<HK, HV> get(K key);
+
+   boolean remove(K key);
+
+   <HK> boolean remove(K key, HK hashKey, HK ... others);
+
+   boolean exists(K key);
+
+   <HK> boolean containsKey(K key, HK hashKey);
+
+   <HK, HV> Hash<HK, HV> submit(Hash<HK, HV> hash); // <2>
+}
+----
+
+<1> Returning a `Map<HK, HV>` to contain all the vargs keys.
+<2> The `submit` method is used to apply the local changes to the server's Hash. The server's `Hash<?, ?>` is returned.
```

Comment:
Do we really need the return value? Hash could be very big. Returning on request or only if submit fails?

Reply:
Yeah, but not strictly. I'll add some void methods, too. In one case, for example, the RESP connector needs to return what has changed. But I was aiming at something more general instead of focusing only on the RESP needs.